oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -117,7 +98,83 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      * Validates access token and refreshes it if necessary.
      */
     @Override
-    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+
+        tokenFuture.handle((token, ignored) -> {
+            if (isValidToken(token)) {
+                future.complete(token);
+                return null;
+            }
+            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+                if (token == null && loadTokenFunc != null) {
+                    loadTokenFunc.get().handle((storedToken, unused) -> {
+                        if (isValidToken(storedToken)) {
+                            future.complete(storedToken);
+                            return null;
+                        }
+                        issueAccessToken(token, future);
+                        return null;
+                    });
+                    return null;
+                }
+                if (token != null && token.isRefreshable()) {
+                    refreshAccessToken(token, future);
+                    return null;
+                }
+                issueAccessToken(token, future);
+                return null;
+            }
+            this.tokenFuture.handle((newToken, cause) -> {
+                if (cause != null) {
+                    future.completeExceptionally(cause);
+                } else {
+                    future.complete(newToken);
+                }
+                return null;
+            });
+            return null;
+        });
+        return future;
+    }
```

Comment:
`tokenFuture` is initialized with an empty future which is useless.
It also makes the code slanted to the right because the logic starts in `handle` function.
So how about `tokenFuture` as `Nullable` and do this?
```suggestion
    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
        GrantedOAuth2AccessToken token = null;
        if (tokenFuture != null && !tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }

        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
        if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
            this.tokenFuture.handle((newToken, cause) -> {
                if (cause != null) {
                    future.completeExceptionally(cause);
                } else {
                    future.complete(newToken);
                }
                return null;
            });
            return future;
        }

        if (token == null && loadTokenFunc != null) {
            loadTokenFunc.get().handle((storedToken, unused) -> {
                if (isValidToken(storedToken)) {
                    future.complete(storedToken);
                    return null;
                }
                issueAccessToken(null, future);
                return null;
            });
            return future;
        }

        if (token != null && token.isRefreshable()) {
            refreshAccessToken(token, future);
            return future;
        }
        issueAccessToken(token, future);
        return future;
    }
```

Reply:
Wouldn't this code block the thread somehow? I don't think `tokenFuture` is not guaranteed to have completed at the time we call `join()`.
```java
if (tokenFuture != null && !tokenFuture.isCompletedExceptionally()) {
    token = tokenFuture.join();
```

Reply:
That's correct. 😅 
We need to do:
```java
if (tokenFuture != null && tokenFuture.isDone() && !tokenFuture.isCompletedExceptionally()) {
```

Reply:
Even after adding `tokenFuture.isDone()` and making `tokenFuture` `@Nullable`, it doesn't pass [these unit tests](https://github.com/line/armeria/pull/3618/files#diff-5e194929dbd201eac45e8aaa05dce13942127cd33ba8632df3a7921aa21549aeR177-R238) for the reason being...:
* Let's say `#1 getAccessToken()` call atomically updated `tokenFuture` and it takes 200ms for the future to complete.
* `#2 getAccessToken()` comes in at `t = 10ms` and `tokenFuture.isDone()` evaluates to `false` and it tries to update `tokenFuture` and initiate another token issuing flow.
* Updating `tokenFuture` by `#2` will still succeed even though the future updated in `#1 getAccessToken()` is still pending. Expected behavior of `#2` is to wait for the `tokenFuture`, i guess?

Reply:
Ah-ha, that's correct. 😅 Let me update the code.

Reply:
Could you try with this, please?
```java
public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
    final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
    GrantedOAuth2AccessToken token = null;
    if (tokenFuture != null) {
        if (!tokenFuture.isDone()) {
            return tokenFuture;
        }
        if (!tokenFuture.isCompletedExceptionally()) {
            token = tokenFuture.join();
            if (isValidToken(token)) {
                return tokenFuture;
            }
        }
    }

    final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
    if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
        return this.tokenFuture;
    }
    if (token == null && loadTokenFunc != null) {
        loadTokenFunc.get().handle((storedToken, unused) -> {
            if (isValidToken(storedToken)) {
                future.complete(storedToken);
                return null;
            }
            issueAccessToken(null, future);
            return null;
        });
        return future;
    }
    if (token != null && token.isRefreshable()) {
        refreshAccessToken(token, future);
        return future;
    }
    issueAccessToken(token, future);
    return future;
}

```

Reply:
Now I get it, thanks! This works like a charm!
Just one thing. What about making `this.tokenFuture` `@Nonnull`? This will still work and we can remove null check.

Reply:
Glad that it's working now. I should've tested it before I paste the code. 😅 

> What about making this.tokenFuture @Nonnull? This will still work and we can remove null check.

The code I pasted is not an answer but just a rough idea, so please go ahead to make it better. 😄